gadget/update.go

```diff
@@ -946,6 +948,100 @@ type StructureLocation struct {
+			// partition labels/names are unique in the partition table, we
+			// could do a lookup by matching partition label, but this won't
+			// work on MBR which doesn't have such a concept, so instead we
+			// use the start offset to locate which disk partition this
+			// structure is equal to.
+			partitions, err := disk.Partitions()
+			if err != nil {
+				return nil, err
+			}
+
+			var foundP disks.Partition
+			found := false
+			for _, p := range partitions {
+				if p.StartInBytes == uint64(structStartOffset) {
+					foundP = p
+					found = true
+					break
+				}
+			}
+			if !found {
+				return nil, fmt.Errorf("cannot locate structure %d on volume %s: no matching start offset", volYamlIndex, vol.Name)
+			}
+
+			// if this structure is an encrypted one, then we can't just
+			// get the root mount points for the device node, we would need
+			// to find the decrypted mapper device for the encrypted device
+			// node and then find the root mount point of the mapper device
+			if _, ok := encryptionParams[volStruct.Name]; ok {
+				logger.Noticef("gadget asset update for assets on encrypted partition %s unsupported", volStruct.Name)
+
+				// leaving this structure as an empty location will
+				// mean when an update to this structure is actually
+				// performed it will fail, but we won't fail updates to
+				// other structures - it is treated like an unmounted
+				// partition
+				locations[volYamlIndex] = loc
+				continue
+			}
+
+			// otherwise normal unencrypted filesystem, find the rw mount
+			// points
+			mountpts, err := disks.MountPointsForPartitionRoot(foundP, map[string]string{"rw": ""})
+			if err != nil {
+				return nil, fmt.Errorf("cannot locate structure %d on volume %s: error searching for root mount points: %v", volYamlIndex, vol.Name, err)
+			}
+			var mountpt string
+			if len(mountpts) == 0 {
+				// this filesystem is not already mounted, we probably
+				// should mount it in order to proceed with the update?
+
+				// TODO: do something better here?
+				logger.Noticef("structure %d on volume %s (%s) is not mounted read/write anywhere to be able to update it", volYamlIndex, vol.Name, foundP.KernelDeviceNode)
+			} else {
+				// use the first one, it doesn't really matter to us
+				// which one is used to update the contents
+				mountpt = mountpts[0]
+			}
+			loc.RootMountPoint = mountpt
+		} else {
+			// no filesystem, the device for this one is just the device
+			// for the disk itself
+			loc.Device = disk.KernelDeviceNode()
+			loc.Offset = structStartOffset
+
+			// Specifically for eMMC devices, the boot0 and boot1 partitions are not
+			// really partitions, but actually pseudo devices. So even though there
+			// is no filesystem, we still must address each sub-device like if the
+			// device had a filesystem
+			if vol.Schema == schemaEMMC {
+				switch volStruct.Name {
+				case "boot0", "boot1":
+					loc.Device += volStruct.Name
```

Comment:
would it make sense to extend StructureLocation with additional field eg. ParentDevice which is set only if Device is some sort of virtual device owned by the parent but not directly mapped mapped through its content? In which case the regex and name guessing in Update/Rollback wouldn't be needed

Reply:
I'm not sure this will clear up the special handling, say we make a `ParentDevice` == /dev/mmcblk0, and `Device` == /dev/mmcblk0boot0, we will still need to extract `mmcblk0boot0` from the string. Granted if the device always appear as a `/dev/...` then it's easy to just remove the prefix, but we would still need to do that to get the device name itself and use that in sysfs, whether we have a `ParentDevice`.

Also, does it even make sense to engineer a generic solution when we have no other use-cases for this?